File: frontend/src/utils/passwordValidation.ts
export interface PasswordValidation {
  isValid: boolean;
  errors: string[];
  strength: 'weak' | 'medium' | 'strong';
}

/**
 * Validates password strength
 * @param password - The password to validate
 * @returns Validation result with errors and strength indicator
 */
export const validatePassword = (password: string): PasswordValidation => {
  const errors: string[] = [];
  
  // Minimum length check
  if (password.length < 8) {
    errors.push('Must be at least 8 characters long');
  }
  
  // Uppercase check
  if (!/[A-Z]/.test(password)) {
    errors.push('Must contain at least one uppercase letter');
  }
  
  // Lowercase check
  if (!/[a-z]/.test(password)) {
    errors.push('Must contain at least one lowercase letter');
  }
  
  // Number check
  if (!/[0-9]/.test(password)) {
    errors.push('Must contain at least one number');
  }
  
  // Special character check (optional but recommended)
  // if (!/[!@#$%^&*()_+\-=\[\]{};':"\\|,.<>\/?]/.test(password)) {
  //   errors.push('Must contain at least one special character');
  // }
  
  // Calculate strength based on number of missing requirements
  const missingChecks = errors.length;
  let strength: 'weak' | 'medium' | 'strong' = 'strong';
  
  if (missingChecks >= 3) {
    strength = 'weak';
  } else if (missingChecks >= 1) {
    strength = 'medium';
  }
  
  return {
    isValid: errors.length === 0,
    errors,
    strength
  };
};

File: backend/app/utils/password_validation.py
import re
from typing import Tuple, List

def validate_password(password: str) -> Tuple[bool, List[str]]:
    """
    Validates password strength.
    Returns (is_valid, errors) where errors is a list of error messages.
    """
    errors = []
    
    # Minimum length check
    if len(password) < 8:
        errors.append("Password must be at least 8 characters long")
    
    # Uppercase check
    if not re.search(r'[A-Z]', password):
        errors.append("Password must contain at least one uppercase letter")
    
    # Lowercase check
    if not re.search(r'[a-z]', password):
        errors.append("Password must contain at least one lowercase letter")
    
    # Number check
    if not re.search(r'[0-9]', password):
        errors.append("Password must contain at least one number")
    
    # Special character check (optional)
    # if not re.search(r'[!@#$%^&*()_+\-=\[\]{};':"\\|,.<>\/?]', password):
    #     errors.append("Password must contain at least one special character")
    
    return len(errors) == 0, errors